core/src/transaction_scheduler.rs

```diff
@@ -0,0 +1,1631 @@
+            deserialized_packet.immutable_section(),
+            &bank.feature_set,
+            bank.vote_only_bank(),
+            bank.as_ref(),
+        )
+        .ok_or_else(|| SchedulerError::InvalidSanitizedTransaction)?;
+
+        let priority = deserialized_packet.immutable_section().priority();
+
+        {
+            let mut scheduled_accounts_l = scheduled_accounts.lock().unwrap();
+
+            let account_locks = sanitized_tx
+                .get_account_locks(&bank.feature_set)
+                .map_err(|e| SchedulerError::InvalidTransactionFormat(e))?;
+
+            trace!(
+                "popped tx w/ priority: {}, readable: {:?}, writeable: {:?}",
+                priority,
+                account_locks.readonly,
+                account_locks.writable
+            );
+
+            // Make sure that this transaction isn't blocked on another transaction that has a higher
+            // fee for one of its accounts
+            if let Err(e) = Self::check_accounts_not_blocked(
+                &scheduled_accounts_l,
+                highest_wl_blocked_account_fees,
+                highest_rl_blocked_account_fees,
+                &account_locks.writable,
+                &account_locks.readonly,
+            ) {
+                Self::upsert_higher_fee_account_lock(
+                    &account_locks.writable,
+                    &account_locks.readonly,
+                    highest_wl_blocked_account_fees,
+                    highest_rl_blocked_account_fees,
+                    priority,
+                );
+                return Err(e);
+            }
+
+            // TODO what else does banking stage do to filter transactions? should do that here
+            //  so batch is guaranteed to be ok
+            // - run QoS (cost model)
+            // - lock accounts with QoS results
+            // - check blockhash age
+            // - check to make sure it's not already processed
+            // - check fee payer has a high enough balance
+            //
+            // if cost model is exceeded, throw it in another buffer
+
+            // already checked we can lock accounts in check_accounts_not_blocked
+            Self::lock_accounts(
+                &mut scheduled_accounts_l,
+                &account_locks.writable,
+                &account_locks.readonly,
+            )
+            .unwrap();
+
+            Ok(sanitized_tx)
+        }
+    }
+
+    fn upsert_higher_fee_account_lock(
+        writable_keys: &[&Pubkey],
+        readonly_keys: &[&Pubkey],
+        highest_wl_blocked_account_fees: &mut HashMap<Pubkey, u64>,
+        highest_rl_blocked_account_fees: &mut HashMap<Pubkey, u64>,
+        priority: u64,
+    ) {
+        for acc in readonly_keys {
+            match highest_rl_blocked_account_fees.entry(**acc) {
+                Entry::Occupied(mut e) => {
+                    if priority > *e.get() {
+                        // NOTE: this should never be the case!
```

Comment:
just to make sure I understand correctly, this should never be the case because `get_scheduled_batch` is popping off max priority txs and these maps don't outlive that function? i.e. the earlier iterations have already inserted higher priority to pubkey mappings

Reply:
Yeah!